pkg/bbgo/margin_info_updater.go

```diff
@@ -0,0 +1,91 @@
+package bbgo
+
+import (
+	"context"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/c9s/bbgo/pkg/util/timejitter"
+)
+
+type MaxBorrowableCallbackType func(asset string, amount fixedpoint.Value)
+
+//go:generate callbackgen -type MarginInfoUpdater
+type marginInfoUpdaterType struct {
+	BorrowRepayService types.MarginBorrowRepayService
+
+	assets                 map[string]struct{}
+	maxBorrowableCallbacks []MaxBorrowableCallbackType
+}
+
+func NewMarginInfoUpdaterFromSession(
+	session *ExchangeSession,
```

Comment:
you only need session.Exchange for this MarginInfoUpdater, the margin flag checking can be done outside of this component

Reply:
The symantic for `NewMarginInfoUpdaterFromSession` is that it is the constructor of the private type `marginInfoUpdaterType`.
It will try to construct the struct and return `nil` if it's not viable.

The caller only need to check if the value returned is `nil` or not.

Reply:
ok, fixed.